src/main/test/ApplicationUtilsTests.cpp

```diff
@@ -146,3 +150,197 @@ TEST_CASE("offline self-check works", "[applicationutils][selfcheck]")
+
+        simulation->crankUntil(
+            [&]() { return simulation->haveAllExternalized(3, 5); },
+            10 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+        // Generate a bit of load, and crank for some time
+        auto& loadGen = validator->getLoadGenerator();
+        loadGen.generateLoad(true, /* nAccounts */ 10, 0, 0, /*txRate*/ 1,
+                             /*batchSize*/ 1, std::chrono::seconds(0), 0);
+
+        auto& loadGenDone = validator->getMetrics().NewMeter(
+            {"loadgen", "run", "complete"}, "run");
+        auto currLoadGenCount = loadGenDone.count();
+
+        auto checkpoint =
+            validator->getHistoryManager().getCheckpointFrequency();
+
+        uint32_t selectedLedger = 0;
+        std::string selectedHash;
+
+        // Make sure validator publishes something
+        simulation->crankUntil(
+            [&]() {
+                bool loadDone = loadGenDone.count() > currLoadGenCount;
+                auto lcl =
+                    validator->getLedgerManager().getLastClosedLedgerHeader();
+                // Pick some ledger in the second checkpoint to run catchup
+                // against later
+                if (lcl.header.ledgerSeq < 2 * checkpoint)
+                {
+                    selectedLedger = lcl.header.ledgerSeq;
+                    selectedHash = binToHex(lcl.hash);
+                }
+
+                // Validator should publish at least two checkpoints
+                return loadDone &&
+                       simulation->haveAllExternalized(3 * checkpoint, 5);
+            },
+            50 * Herder::EXP_LEDGER_TIMESPAN_SECONDS, false);
+
+        auto lcl = simulation->getNode(vNode2NodeID)
+                       ->getLedgerManager()
+                       .getLastClosedLedgerHeader();
+
+        // Shutdown captive core and test various startup paths
+        simulation->removeNode(vNode2NodeID);
+
+        SECTION("minimal DB setup")
+        {
+            // Shutdown captive core
+            SECTION("not found")
+            {
+                // Remove `buckets` dir completely
+                fs::deltree(cfg2.BUCKET_DIR_PATH);
+
+                // Initialize new minimal DB from scratch
+                auto app = setupApp(cfg2, clock, 0, "");
+                REQUIRE(app);
+                REQUIRE(checkState(*app));
+            }
+            SECTION("found")
+            {
+                // Found existing minimal DB, reset to genesis
+                auto app = setupApp(cfg2, clock, 0, "");
+                REQUIRE(app);
+                REQUIRE(checkState(*app));
+            }
+        }
+        SECTION("rebuild state")
+        {
+            SECTION("from buckets")
+            {
+                selectedLedger = lcl.header.ledgerSeq;
+                selectedHash = binToHex(lcl.hash);
+
+                auto app = setupApp(cfg2, clock, selectedLedger, selectedHash);
+                REQUIRE(app);
```

Comment:
Maybe require no-replay happened here?

Reply:
Added a combination of checks: `canRebuildInMemoryLedgerFromBuckets` and the state of LedgerManager (if catchup indeed happened, LM is in LM_CATCHING_UP_STATE state at the end)